src/stream/src/common/log_store/kv_log_store/reader.rs

```diff
@@ -0,0 +1,246 @@
+use risingwave_storage::hummock::CachePolicy;
+use risingwave_storage::store::{PrefetchOptions, ReadOptions};
+use risingwave_storage::StateStore;
+use tokio_stream::StreamExt;
+
+use crate::common::log_store::kv_log_store::buffer::{LogStoreBufferItem, LogStoreBufferReceiver};
+use crate::common::log_store::kv_log_store::serde::{
+    new_log_store_item_stream, LogStoreItemStream, LogStoreRowSerde,
+};
+use crate::common::log_store::{LogReader, LogStoreError, LogStoreReadItem, LogStoreResult};
+
+enum ReaderState<S: StateStore> {
+    /// No data has been read yet
+    Uninitialized,
+    /// Consuming data previously written to state store before the reader was initialized.
+    ConsumingStateStore {
+        first_write_epoch: u64,
+        state_store_stream: Pin<Box<LogStoreItemStream<S::IterStream>>>,
+    },
+    /// Consuming newly written data after the reader was initialized.
+    ConsumingStream { epoch: u64 },
+}
+
+pub struct KvLogStoreReader<S: StateStore> {
+    table_id: TableId,
+
+    state_store: S,
+
+    serde: LogStoreRowSerde,
+
+    rx: LogStoreBufferReceiver,
+
+    reader_state: ReaderState<S>,
+
+    latest_barrier: Option<u64>,
+}
+
+impl<S: StateStore> KvLogStoreReader<S> {
+    pub(crate) fn new(
+        table_id: TableId,
+        state_store: S,
+        serde: LogStoreRowSerde,
+        rx: LogStoreBufferReceiver,
+    ) -> Self {
+        Self {
+            table_id,
+            state_store,
+            reader_state: ReaderState::Uninitialized,
+            serde,
+            rx,
+            latest_barrier: None,
+        }
+    }
+
+    fn recv_barrier(&mut self, epoch: u64) {
+        if let Some(prev_epoch) = &self.latest_barrier {
+            assert!(
+                epoch > *prev_epoch,
+                "next epoch {} should be greater than prev epoch {}",
+                epoch,
+                prev_epoch
+            );
+        }
+        self.latest_barrier = Some(epoch);
+    }
+}
+
+impl<S: StateStore> LogReader for KvLogStoreReader<S> {
+    type InitFuture<'a> = impl Future<Output = LogStoreResult<()>> + 'a;
+    type NextItemFuture<'a> = impl Future<Output = LogStoreResult<(u64, LogStoreReadItem)>> + 'a;
+    type TruncateFuture<'a> = impl Future<Output = LogStoreResult<()>> + 'a;
+
+    fn init(&mut self) -> Self::InitFuture<'_> {
+        async move {
+            let first_write_epoch = self.rx.init().await;
+            let streams = try_join_all(self.serde.vnodes().iter_vnodes().map(|vnode| {
+                let range_start = Bytes::from(Vec::from(vnode.to_be_bytes()));
+                let range_end = self.serde.serialize_epoch(vnode, first_write_epoch);
+                let table_id = self.table_id;
+                let state_store = self.state_store.clone();
+                async move {
+                    state_store
+                        .iter(
+                            (Included(range_start), Excluded(range_end)),
+                            u64::MAX,
```

Comment:
We have a `safe_epoch` in hummock version. Reading below the safe epoch may return an error even though it is correct. Since we don't pin the epoch in log reader, the safe epoch will bump up to above the first write epoch, and then read from the first write epoch may return an error.